Add tests for stage 3 proposal "private fields in"

jugglinmike opened the pull request:
I have two questions about the "private class fields" proposal which have bearing on the tests for this proposal:

- [the fallibility of ResolveBinding when applied to a PrivateEnvironment](https://github.com/tc39/ecma262/pull/1668/files#r584007453)
- [the expected behavior of GetValue when applied to a private name binding](https://github.com/tc39/ecma262/pull/1668/files#r570729429)

ljharb commented on test/language/expressions/in/private-field-invalid-identifier-simple.js:
```diff
@@ -0,0 +1,25 @@
+
+$DONOTEVALUATE();
+
+#name in {};
```
confirming i understand negative parse tests: this is ensuring that this is always a syntax error?

rwaldron replied:
Correct!

ljharb commented on test/language/expressions/in/private-field-invalid-rhs.js:
```diff
@@ -0,0 +1,26 @@
+  #field;
+
+  constructor() {
+    #field in () => {};
```
this seems like it should parse just fine, but the test seems to be a negative one?

jugglinmike replied:
I assumed that would be redundant, but it couldn't hurt!

jugglinmike replied:
This doesn't parse because the nonterminal following the `in` keyword is ShiftExpression, but [ArrowFunction is an AssignmentExpression](https://tc39.es/ecma262/#sec-expressions). That's the same reason `yield` has to be parenthesized.

jugglinmike replied:
We could add one, but between the test for YieldExpression (which uses ParenthesizedExpression) and the test for ShiftExpression (which does not), that seems more like a test for the wider grammar than for this proposal specifically. It makes me wonder if anyone has looked into precisely verifying expression precedence...

ljharb commented on test/language/expressions/in/private-field-presence-accessor.js:
```diff
@@ -0,0 +1,46 @@
+         then return true.
+  11. Return false.
+esid: sec-relational-operators-runtime-semantics-evaluation
+features: [class-fields-private-in]
```
does this also need private class fields themselves?

jugglinmike replied:
I assumed that would be redundant, but it couldn't hurt!

ljharb commented on test/language/expressions/in/private-field-rhs-yield-absent.js:
```diff
@@ -0,0 +1,33 @@
+  #field;
+
+  static method() {
+    #field in yield;
```
#foundpoetry

jugglinmike replied:
I'd be happy to add a `shield:` label if we can think of any way that might possibly improve coverage.

jugglinmike replied:
`keeled: healed = #field in yield shield(peeled)`

ljharb commented on test/language/expressions/in/private-field-presence-method.js:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  static init() {
+    Child = class {
```
might be nice to also have a few of these tests where the child doesn’t shadow the field

jugglinmike replied:
Will do

ljharb reviewed (approved):
Thank you, this seems great to me!

jugglinmike commented:
@ljharb Sorry to bug you, but would you mind approving this again? I had to fix a typo in the features flags, and that cleared your review.

rwaldron commented:
@jugglinmike please rebase and then I will merge this! Thanks!

jugglinmike commented:
Right on, @rwaldron. I've rebased, squashed, and force-pushed the result. [The prior version of this branch is available here](https://github.com/bocoup/test262/pull/new/private-fields-in-orig) in case that's of interest to anyone.

jugglinmike commented:
Strange. I force-pushed to this Pull Request's branch, and although the new commit is visible from the source repository's page, it's not reflected here.

jugglinmike commented:
I re-created the commit and force-pushed a second time, and the result is (finally) visible here. Ready when you are, @rwaldron

leobalter commented:
I'm merging this - after a proper review - to avoid extending more time with this PR open. The tests look great and thanks for holding and cleaning up the rebased commit, @jugglinmike!
